Record types (#6679)
nextflow-io/nextflow: docs/tutorials/records.md

(migrating-records)=

# Migrating to records

Nextflow 26.04 introduces the ability to use *records* and *record types* in a Nextflow pipeline. This tutorial demonstrates how to migrate to records using the [rnaseq-nf](https://github.com/nextflow-io/rnaseq-nf) pipeline as an example.

## Overview

Nextflow pipelines need a way to model composite data, such as a *sample* consisting of a unique identifer, some files, and other metadata.

The standard way to model composite data in Nextflow is with a *tuple*:

```nextflow
sample = tuple([id: '1'], file('1_1.fastq'), file('1_2.fastq'))

channel.of(sample).view { meta, fastq_1, fastq_2 ->
    "id: ${meta.id}, fastq_1: ${fastq_1.name}, fastq_2: ${fastq_2.name}"
}
```

> Tuples are easy to use, but have several downsides:

Reviewer: ```suggestion
Tuples are easy to use, but have several limitations:
```
Suggestion only

- **Brittle data structures**: Tuples must be repeatedly unpacked and repacked whenever they pass through a channel operator. The tuple can instead be passed as a single argument, in which case tuple elements must be accessed by index. Either way, the user has to remember the names and ordering of the tuple elements.

- **Tight coupling between processes and workflow logic**: When calling a process, the tuple elements must match the process inputs with the same ordering. Getting this wrong can lead to runtime errors that are hard to debug.

- **Limited type checking**: The type checker can validate whether an incoming tuple has the right number of elements and the right element types, but it cannot validate the semantic name of each element, because tuple elements don't have names. For example, the type `Tuple<String,Path,Path>` does not tell you what any of the elements refer to, only their types. As a result, there are many kinds of semantic errors that cannot be caught by the type checker.

Records are a new data type for modelling complex data structures, as an alternative to tuples:

> - **Meaningful data structures**: Record fields are accessed by name instead of index, allowing you to think about your data in terms of meaningful names instead of arbitrary indices.

Reviewer: ```suggestion
- **Meaningful data structures**: Record fields are accessed by name instead of index. They allow you to think about your data in terms of meaningful names instead of arbitrary indices.
```

:::

### Migrating pipeline inputs

The rnaseq-nf pipeline takes a CSV file, also known as a *samplesheet*, containing a row for each input sample. This samplesheet is loaded into a channel of tuples:

```nextflow
read_pairs_ch = channel.of(params.reads)
    .flatMap { csv -> csv.splitCsv() }
    .map { row ->
        tuple(row[0], file(row[1]), file(row[2]))
    }
```

To load the samplesheet as a channel of records, replace the `tuple()` call with `record()` as follows:

```nextflow
read_pairs_ch = channel.of(params.reads)
    .flatMap { csv -> csv.splitCsv() }
    .map { row ->
        record(id: row[0], fastq_1: file(row[1]), fastq_2: file(row[2]))
    }
```

### Migrating workflows

The `RNASEQ` workflow has the following definition after converting it to static types:

```nextflow
workflow RNASEQ {
    take:
    reads_ch        : Channel<Tuple<String,Path,Path>>
    transcriptome   : Path

    main:
    index = INDEX(transcriptome)
    fastqc_ch = FASTQC(reads_ch)
    quant_ch = QUANT(reads_ch, index)
    samples_ch = fastqc_ch.join(quant_ch)

    emit:
    samples : Channel<Tuple<String,Path,Path>> = samples_ch
    index   : Value<Path> = index
}
```

> The `reads_ch` input needs to be refactored as a channel of records. While we could use `Channel<Record>`, it is better to use a *record type* that specifies the required fields. To do this, we need to examine the processes that use `ch_reads` and construct a record type that encapsulates those uses.

Reviewer: ```suggestion
The `reads_ch` input needs to be refactored as a channel of records. While you could use `Channel<Record>`, it is better to use a *record type* that specifies the required fields. To do this, you need to examine the processes that use `ch_reads` and construct a record type that encapsulates those uses.
```
